lib/manager/regex/index.ts

```diff
@@ -1,55 +1,161 @@
-              'Error compiling template for custom manager'
-            );
-            return null;
-          }
-        } else if (groups[field]) {
-          dep[field] = groups[field];
+      matches.push(matchResult);
+    }
+  } while (matchResult);
+  return matches;
+}
+
+function createDependency(
+  matchResult: RegExpMatchArray,
+  config: CustomExtractConfig,
+  dep?: PackageDependency
+): PackageDependency {
+  const dependency = dep || {};
+  const { groups } = matchResult;
+  for (const field of fields) {
+    const fieldTemplate = `${field}Template`;
+    if (config[fieldTemplate]) {
+      try {
+        dependency[field] = template.compile(config[fieldTemplate], groups);
+      } catch (err) {
+        logger.warn(
+          { template: config[fieldTemplate] },
+          'Error compiling template for custom manager'
+        );
+        return null;
+      }
+    } else if (groups[field]) {
+      dependency[field] = groups[field];
+    }
+  }
+  dependency.replaceString = String(matchResult[0]);
+  return dependency;
+}
+
+function mergeDependency(deps: PackageDependency[]): PackageDependency {
+  const result: PackageDependency = {};
+  deps.forEach((dep) => {
+    fields.forEach((field) => {
+      if (dep[field]) {
+        result[field] = dep[field];
+        // save the line replaceString of the section which contains the current Value for a speed up lookup during the replace phase
+        if (field === 'currentValue') {
+          result.replaceString = dep.replaceString;
         }
       }
-      dep.replaceString = String(matchResult[0]);
-      deps.push(dep);
+    });
+  });
+  return result;
+}
+
+function handleAny(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig
+): PackageDependency[] {
+  return config.matchStrings
+    .map((matchString) => regEx(matchString, 'g'))
+    .flatMap((regex) => regexMatchAll(regex, content)) // match all regex to content, get all matches, reduce to single array
+    .map((matchResult) => createDependency(matchResult, config));
+}
+
+function handleCombination(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig
+): PackageDependency[] {
+  const dep = handleAny(content, packageFile, config).reduce(
+    (mergedDep, currentDep) => mergeDependency([mergedDep, currentDep]),
+    {}
+  ); // merge fields of dependencies
+  return [dep];
+}
+
+function handleRecursive(
+  content: string,
+  packageFile: string,
+  config: CustomExtractConfig,
+  index = 0
+): PackageDependency[] {
+  const regexes = config.matchStrings.map((matchString) =>
+    regEx(matchString, 'g')
+  );
+  // abort if we have no matchString anymore
+  if (regexes[index] == null) {
+    return [];
+  }
+  return regexMatchAll(regexes[index], content).flatMap((match) => {
+    // if we have a depName and a currentValue with have the minimal viable definition
+    if (match?.groups?.depName && match?.groups?.currentValue) {
```

Comment:
Why would we stop matching if there are still regexes left? I don't understand this scenario

Reply:
I have thought of a scenario where the dependency could in one of the levels of the recursion.     
In that case you can define the group values on both levels.    
If the earlier match fails, it will continue to the next level. 

e.g. 
``` json 
{
  "group1": {
    "name": "prom/prometheus",
    "type": "docker",
    "value": "v2.19.0"
  },
  "group2": {
    "test": {
      "name": "grafana/loki",
      "type": "docker",
      "value": "1.6.1"
    },
    "test2": {
      "name": "grafana/loki",
      "type": "docker",
      "value": "1.6.1"
    }
  },
}
```

But in the most cases the user will setup the group values on the last `matchStrings`.